Source/Core/VideoBackends/D3D12/D3DTexture.cpp

```diff
@@ -0,0 +1,316 @@
+// Copyright 2010 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Common/CommonTypes.h"
+#include "Common/MsgHandler.h"
+#include "VideoBackends/D3D12/D3DBase.h"
+#include "VideoBackends/D3D12/D3DCommandListManager.h"
+#include "VideoBackends/D3D12/D3DDescriptorHeapManager.h"
+#include "VideoBackends/D3D12/D3DTexture.h"
+#include "VideoBackends/D3D12/D3DUtil.h"
+#include "VideoBackends/D3D12/FramebufferManager.h"
+#include "VideoBackends/D3D12/Render.h"
+
+namespace DX12
+{
+
+namespace D3D
+{
+
+static ID3D12Resource* s_texture_upload_heaps[2] = {};
+void* s_texture_upload_heap_data_pointers[2] = {};
+
+unsigned int s_texture_upload_heap_current_index = 0;
+unsigned int s_texture_upload_heap_current_offset = 0;
+
+static unsigned int s_texture_upload_heap_size = 8 * 1024 * 1024;
+
+void MoveToNextD3DTextureUploadHeap()
+{
+	s_texture_upload_heap_current_index =
+		(s_texture_upload_heap_current_index + 1) % ARRAYSIZE(s_texture_upload_heaps);
+
+	s_texture_upload_heap_current_offset = 0;
+}
+
+void CleanupPersistentD3DTextureResources()
+{
+	// All GPU work has stopped at this point, can destroy immediately.
+
+	for (UINT i = 0; i < ARRAYSIZE(s_texture_upload_heaps); i++)
+	{
+		SAFE_RELEASE(s_texture_upload_heaps[i]);
+		s_texture_upload_heap_data_pointers[i] = nullptr;
+	}
+
+	s_texture_upload_heap_current_index = 0;
+	s_texture_upload_heap_current_offset = 0;
+}
+
+void ReplaceRGBATexture2D(ID3D12Resource* texture12, const u8* buffer, unsigned int width, unsigned int height, unsigned int src_pitch, unsigned int level, D3D12_RESOURCE_STATES current_resource_state)
+{
+	unsigned int upload_size =
```

Comment:
W/r/t GetCopyableFootprints - that would be a little cleaner, but I'd rather avoid the cost if so. In response to this, I made a dedicated AlignValue function, instead of having this alignment code sprinkled throughout, so hopefully it is more readable now.

I'll address the memory allocation tracking in the next commit...

Using a dedicated copy queue for texture data upload would be a good optimization, after this PR I'll try to profile the potential win there.